src/Analyzer/Passes/LikePerfectAffixRewritePass.cpp

```diff
@@ -0,0 +1,126 @@
+#include <Analyzer/Passes/LikePerfectAffixRewritePass.h>
+
+#include <Analyzer/ConstantNode.h>
+#include <Analyzer/FunctionNode.h>
+#include <Analyzer/IQueryTreeNode.h>
+#include <Analyzer/IdentifierNode.h>
+#include <Analyzer/InDepthQueryTreeVisitor.h>
+#include <Core/Settings.h>
+#include <Common/StringUtils.h>
+#include <DataTypes/IDataType.h>
+#include <Functions/FunctionFactory.h>
+
+#include <boost/algorithm/string.hpp>
+
+namespace DB
+{
+namespace Setting
+{
+    extern const SettingsBool optimize_rewrite_like_perfect_affix;
+}
+
+namespace
+{
+
+/** Rewrite LIKE expressions with perfect affix (prefix or suffix) to startsWith or endsWith functions.
+  * The scope of the rewrite:
+  *     - `col LIKE 'Prefix%'`: as `startsWith(col, 'Prefix')`
+  *     - `col LIKE '%Suffix'`: as `endsWith(col, 'Suffix')`
+  *     - `NOT LIKE `: with `NOT`
+  * Out of scope:
+  *     - `ILIKE` is not rewritten until starts/endsWithCaseInsensitive is implemented
+  * Type requirement on the left operand, `col`:
+  *     - `startsWith`: Only resolved column of type `String` or `FixedString`
+  *     - `endsWith`: Only resolved column of type `String`
+  *     - `LowCardinality(String)` is unsupported due to non-order-preserving encoding
```

Comment:
I’m not sure why LowCardinality isn’t supported. Also, we can add support for Nullable(String).